Add documentation for connecting to Azure SQL instances with managed identities

lengau opened the pull request:
See issue #6031: https://github.com/sqlalchemy/sqlalchemy/issues/6031

<!-- Provide a general summary of your proposed changes in the Title field above -->

### Description
This adds a subsection to the MSSQL dialect documentation about using Azure credentials to connect to servers.

### Checklist
<!-- go over following points. check them with an `x` if they do apply, (they turn into clickable checkboxes once the PR is submitted, so no need to do everything at once)

-->

This pull request is:

- [x] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

Commit:
Add documentation for connecting to Azure SQL instances with managed identities.

See issue #6031: https://github.com/sqlalchemy/sqlalchemy/issues/6031

zzzeek commented:
as mentioned in #6031 this seems overly complicated and also specific to pyodbc, is this documented on pyodbc's website?     for SQLAlhchemy, special arguments like these are just a passthrough for [connect_args](https://docs.sqlalchemy.org/en/13/core/engines.html#use-the-connect-args-dictionary-parameter) and as this guidance seems extremely specific to pyodbc it seems more of a fit for pyodbc's documentation rather than that of SQLAlchemy.

lengau commented:
In pyodbc itself, you can just say `pyodbc.connect(..., attrs_before={1256: bytes(token_struct})`, as laid out in https://github.com/mkleehammer/pyodbc/issues/228

However, because a sqlalchemy engine can create connections indefinitely into the future, it's necessary to provide a valid token to pyodbc each time we create a connection. Otherwise, the new connection will fail once the initial token has expired. (I was initially using a modified version of the code in the above link, but once my code started running for more than 15 minutes connections started failing.)

Pyodbc doesn't support providing a callable for the value in `attrs_before` so this is the simplest way I've been able to create that solves the problem of long-running engines.

CaselIT commented on lib/sqlalchemy/dialects/mssql/pyodbc.py:
```diff
@@ -79,6 +79,49 @@
+After getting an engine, the credentials need to be sent to ``pyodbc.connect``
+each time a connection is requested. One way to do this is to set up an event
+listener on the engine that adds the credential token to the dialect's connect
+call. The following code snippet allows this:
```
this should end with two colon, so that it's formatted as code. 
tldr `this:` -> `this::`

lengau replied:
Thanks for pointing that out!

Commit:
Fix rST typo

CaselIT commented on lib/sqlalchemy/dialects/mssql/pyodbc.py:
```diff
@@ -79,6 +79,49 @@
+            attrs_before[SQL_COPT_SS_ACCESS_TOKEN] = bytes(token_struct)
+            return dialect.connect(*cargs, **cparams)
+            
+This can then be used as follows:
```
also here `::`

lengau replied:
OK this time I actually properly checked the build and made sure it looks right in the browser :-)

Commit:
The slash is actually necessary

Commit:
Update engine and mssql+pyodbc docs based on suggestions

This implements the suggestions from @zzzeek https://github.com/sqlalchemy/sqlalchemy/issues/6031#issuecomment-797040944

1. Describe generally the way to provide a dynamic token in the engine documentation.
2. Provide a clearer example.

zzzeek commented:
this looks really nice thanks for doing this.  i will pull this through the gerrit process to make sure pep8 still passes and also to backport to 1.3.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision d4bc2a67051ba0886e24d1ee93623780171bb88c of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change d4bc2a67051ba0886e24d1ee93623780171bb88c: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

CaselIT commented on lib/sqlalchemy/dialects/mssql/pyodbc.py:
```diff
@@ -70,15 +69,63 @@
+    <https://docs.microsoft.com/en-us/sql/connect/odbc/using-azure-active-directory#authenticating-with-an-access-token>`_
+    that a connection string when using an access token must not contain 
+    ``UID``, ``PWD``, ``Authentication`` or ``Trusted_Connection`` keywords,
+    a the connection string must be provided directly via ``odbc_connect``.
```
typo `a the`

lengau replied:
Thanks for catching that. Fixed!

Commit:
fix typo

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 00abfffd198a4c0e81bb5d15219dd4e40670ce94 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 00abfffd198a4c0e81bb5d15219dd4e40670ce94 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

sqla-tester commented on lib/sqlalchemy/dialects/mssql/pyodbc.py:
```diff
@@ -70,15 +69,63 @@
+        raw_token = azure_credentials.get_token(TOKEN_URL).token.encode("utf-16-le")
+        token_struct = struct.pack(f"<I{len(raw_token)}s", len(raw_token), token)
+        cparams["attrs_before"] = {SQL_COPT_SS_ACCESS_TOKEN: token_struct}
+        
```
**mike bayer** (zzzeek) wrote:

hey there, can you review what I changed here?  since we are modifying the parameters we can more succinctly get rid of this token in the same method.   if you can make sure this works, this is more succinct.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

sqla-tester reviewed (commented):
**mike bayer** (zzzeek) wrote:

i made a few changes if you can review thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

OK I'm going to merge this as is, if you have time to verify the change I made is OK or not, we can always modify it after the fact.  thanks!

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/2632 has been **merged**. Congratulations! :)
